clients/py/sapphirepy/deoxysii.py

```diff
@@ -0,0 +1,580 @@
+    0x3a3a4e74, 0x0a0a1e14, 0x4949db92, 0x06060a0c, 0x24246c48, 0x5c5ce4b8,
+    0xc2c25d9f, 0xd3d36ebd, 0xacacef43, 0x6262a6c4, 0x9191a839, 0x9595a431,
+    0xe4e437d3, 0x79798bf2, 0xe7e732d5, 0xc8c8438b, 0x3737596e, 0x6d6db7da,
+    0x8d8d8c01, 0xd5d564b1, 0x4e4ed29c, 0xa9a9e049, 0x6c6cb4d8, 0x5656faac,
+    0xf4f407f3, 0xeaea25cf, 0x6565afca, 0x7a7a8ef4, 0xaeaee947, 0x08081810,
+    0xbabad56f, 0x787888f0, 0x25256f4a, 0x2e2e725c, 0x1c1c2438, 0xa6a6f157,
+    0xb4b4c773, 0xc6c65197, 0xe8e823cb, 0xdddd7ca1, 0x74749ce8, 0x1f1f213e,
+    0x4b4bdd96, 0xbdbddc61, 0x8b8b860d, 0x8a8a850f, 0x707090e0, 0x3e3e427c,
+    0xb5b5c471, 0x6666aacc, 0x4848d890, 0x03030506, 0xf6f601f7, 0x0e0e121c,
+    0x6161a3c2, 0x35355f6a, 0x5757f9ae, 0xb9b9d069, 0x86869117, 0xc1c15899,
+    0x1d1d273a, 0x9e9eb927, 0xe1e138d9, 0xf8f813eb, 0x9898b32b, 0x11113322,
+    0x6969bbd2, 0xd9d970a9, 0x8e8e8907, 0x9494a733, 0x9b9bb62d, 0x1e1e223c,
+    0x87879215, 0xe9e920c9, 0xcece4987, 0x5555ffaa, 0x28287850, 0xdfdf7aa5,
+    0x8c8c8f03, 0xa1a1f859, 0x89898009, 0x0d0d171a, 0xbfbfda65, 0xe6e631d7,
+    0x4242c684, 0x6868b8d0, 0x4141c382, 0x9999b029, 0x2d2d775a, 0x0f0f111e,
+    0xb0b0cb7b, 0x5454fca8, 0xbbbbd66d, 0x16163a2c,
+])
+
+def uint8(x:int):
+    return x & 0xFF
+
+def bc_encrypt(ciphertext:bytearray, derived_k:list[bytearray], tweak:bytearray, plaintext:bytes|bytearray):
+    assert len(plaintext) >= 16
+    assert len(derived_k) == STK_COUNT
+    assert len(derived_k[0]) == STK_SIZE
+    assert len(tweak) == TWEAK_SIZE
+
+    # Derive all the Sub-Tweak Keys.
+    stks = derive_sub_tweak_keys(derived_k, tweak)
+
+    # AddRoundTweakKey (AES -> AddRoundKey)
+    s0, s1, s2, s3 = struct.unpack('>LLLL', plaintext[:16])
+    s0 = s0 ^ stks[0][0]
+    s1 = s1 ^ stks[0][1]
+    s2 = s2 ^ stks[0][2]
+    s3 = s3 ^ stks[0][3]
+
+    i = 1
+    while i <= ROUNDS:
+        # SubBytes, ShiftRows, MixBytes (AES -> MixColumns),
+        # AddRoundTweakKey (AES -> AddRoundKey).
+        #
+        # In AES-NI terms, this is *exactly* equivalent to AESENC (though
+        # Intel's notation transposes ShiftRows, and SubBytes).
+        t0 = (TE0[uint8(s0>>24)] ^
+              TE1[uint8(s1>>16)] ^
+              TE2[uint8(s2>>8)] ^
+              TE3[uint8(s3)] ^
+              stks[i][0])
+
+        t1 = (TE0[uint8(s1>>24)] ^
+              TE1[uint8(s2>>16)] ^
+              TE2[uint8(s3>>8)] ^
+              TE3[uint8(s0)] ^
+              stks[i][1])
+
+        t2 = (TE0[uint8(s2>>24)] ^
+              TE1[uint8(s3>>16)] ^
+              TE2[uint8(s0>>8)] ^
+              TE3[uint8(s1)] ^
+              stks[i][2])
+
+        t3 = (TE0[uint8(s3>>24)] ^
+              TE1[uint8(s0>>16)] ^
+              TE2[uint8(s1>>8)] ^
+              TE3[uint8(s2)] ^
+              stks[i][3])
+
+        s0 = t0
+        s1 = t1
+        s2 = t2
+        s3 = t3
+        i += 1
+
+    ciphertext[:16] = struct.pack('>LLLL', s0, s1, s2, s3)
+
+class DeoxysII:
```

Comment:
I hate to be that guy...but should we repo this separately à la Go [package](https://github.com/oasisprotocol/deoxysii)?

Reply:
For now, it is easiest to just include it in the package as it's a single python file that's only used in one place.

We can consider moving it to an external package in the near future